tardis/spectrum/formal_integral/source_function.py

```diff
@@ -0,0 +1,238 @@
+
+        no_lvls = len(levels)
+        no_shells = len(dilution_factor) # TODO: is dilution_factor len=no_of_shells 
+
+        # Calculate e_dot_u
+        upper_level_index = atomic_data.lines.index.droplevel(
+            "level_number_lower"
+        )
+        e_dot_u = calculate_edotu(time_of_simulation, volume, tau_sobolevs, Edotlu_estimator, 
+                        macro_data, macro_ref, self.line_interaction_type,
+                        transition_probabilities, upper_level_index, columns, no_lvls)
+
+
+        # Calculate att_S_ul
+        transition_type = atomic_data.macro_atom_data.transition_type
+        transitions = atomic_data.macro_atom_data[transition_type == -1].copy()
+        transitions_index = transitions.set_index(
+            ["atomic_number", "ion_number", "source_level_number"]
+        ).index.copy()
+        transition_line_id = transitions.transition_line_id.values
+        lines = atomic_data.lines.set_index('line_id') # TODO: investigate why this is like this
+        lines_idx = lines.index.values 
+        
+        att_S_ul = calculate_att_S_ul(lines, transition_probabilities, 
+                            no_of_shells, transition_line_id, lines_idx, 
+                            transitions_index, transition_type,
+                            e_dot_u, time_explosion)
+
+        # Calculate Jredlu and Jbluelu
+        Jbluelu = calculate_Jbluelu(time_explosion, time_of_simulation, volume, j_blue_estimator)
+        Jredlu = calculate_Jredlu(Jbluelu, tau_sobolevs, att_S_ul)
+
+        return SourceFunctionState(att_S_ul, Jredlu, Jbluelu, e_dot_u)
+
+
+# transport
+    # time of sim, Edotlu_estimator, line_interaction_type
+# sim state
+    # volume
+# upper_level_index = atomic_data.lines.index.droplevel("level_number_lower")
+def calculate_edotu(time_of_simulation, volume, tau_sobolevs, Edotlu_estimator, 
+                    macro_data, macro_ref, line_interaction_type,
+                    transition_probabilities, upper_level_idx, columns, no_lvls):
+
+    Edotlu_norm_factor = 1 / (time_of_simulation * volume)
+    exptau = 1 - np.exp(-tau_sobolevs)
+    Edotlu = Edotlu_norm_factor * exptau * Edotlu_estimator
+
+    e_dot_lu = pd.DataFrame(
+        Edotlu.value, index=upper_level_idx, columns=columns
+    )
+    e_dot_u = e_dot_lu.groupby(level=[0, 1, 2]).sum()
+    e_dot_u_src_idx = macro_ref.loc[e_dot_u.index].references_idx.values
+
+    e_dot_u.index.names = [
+        "atomic_number",
+        "ion_number",
+        "source_level_number",
+    ]
+
+    if line_interaction_type == "macroatom":
+        internal_jump_mask = (macro_data.transition_type >= 0).values
+        ma_int_data = macro_data[internal_jump_mask]
+        internal = transition_probabilities[internal_jump_mask]
+
+        source_level_idx = ma_int_data.source_level_idx.values
+        destination_level_idx = ma_int_data.destination_level_idx.values
+
+    # if transport.line_interaction_type == "macroatom":
```

Comment:
You could make `line_interaction_type` a default kwarg passed to this function that the Solver class could then pass when performing calculations